File: app/src/main/java/dev/ragnarok/fenrir/adapter/fave/FaveArticlesAdapter.java
package dev.ragnarok.fenrir.adapter.fave;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import dev.ragnarok.fenrir.Constants;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.model.Article;
import dev.ragnarok.fenrir.model.Photo;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.util.ViewUtils;

public class FaveArticlesAdapter extends RecyclerView.Adapter<FaveArticlesAdapter.Holder> {

    private final Context context;
    private List<Article> data;
    private ClickListener clickListener;

    public FaveArticlesAdapter(List<Article> data, Context context) {
        this.data = data;
        this.context = context;
    }

    @NonNull
    @Override
    public Holder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new Holder(LayoutInflater.from(context).inflate(R.layout.item_article, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull Holder holder, int position) {
        Article article = data.get(position);
        holder.btFave.setImageResource(R.drawable.favorite);
        holder.btFave.setOnClickListener(v -> clickListener.onDelete(position, article));
        if (article.getURL() != null) {
            holder.ivButton.setVisibility(View.VISIBLE);
            holder.ivButton.setOnClickListener(v -> clickListener.onUrlClick(article.getURL()));
        } else
            holder.ivButton.setVisibility(View.GONE);

        holder.btShare.setOnClickListener(v -> clickListener.onShare(article));

        String photo_url = null;
        if (article.getPhoto() != null) {
            photo_url = article.getPhoto().getUrlForSize(Settings.get().main().getPrefPreviewImageSize(), false);
        }

        if (photo_url != null) {
            holder.ivPhoto.setVisibility(View.VISIBLE);
            ViewUtils.displayAvatar(holder.ivPhoto, null, photo_url, Constants.PICASSO_TAG);
            holder.ivPhoto.setOnLongClickListener(v -> {
                clickListener.onPhotosOpen(article.getPhoto());
                return true;
            });
        } else
            holder.ivPhoto.setVisibility(View.GONE);

        if (article.getSubTitle() != null) {
            holder.ivSubTitle.setVisibility(View.VISIBLE);
            holder.ivSubTitle.setText(article.getSubTitle());
        } else
            holder.ivSubTitle.setVisibility(View.GONE);

        if (article.getTitle() != null) {
            holder.ivTitle.setVisibility(View.VISIBLE);
            holder.ivTitle.setText(article.getTitle());
        } else
            holder.ivTitle.setVisibility(View.GONE);

        if (article.getOwnerName() != null) {
            holder.ivName.setVisibility(View.VISIBLE);
            holder.ivName.setText(article.getOwnerName());
        } else
            holder.ivName.setVisibility(View.GONE);
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    public void setData(List<Article> data) {
        this.data = data;
        notifyDataSetChanged();
    }

    public void setClickListener(ClickListener clickListener) {
        this.clickListener = clickListener;
    }

    public interface ClickListener {
        void onUrlClick(String url);

        void onPhotosOpen(Photo photo);

        void onDelete(int index, Article article);

        void onShare(Article article);
    }

    public static class Holder extends RecyclerView.ViewHolder {

        final ImageView ivPhoto;
        final ImageView btFave;
        final ImageView btShare;
        final TextView ivSubTitle;
        final TextView ivTitle;
        final TextView ivName;
        final Button ivButton;

        public Holder(View root) {
            super(root);

            ivPhoto = itemView.findViewById(R.id.item_article_image);
            ivSubTitle = itemView.findViewById(R.id.item_article_subtitle);
            ivTitle = itemView.findViewById(R.id.item_article_title);
            ivName = itemView.findViewById(R.id.item_article_name);
            ivButton = itemView.findViewById(R.id.item_article_read);
            btFave = itemView.findViewById(R.id.item_article_to_fave);
            btShare = itemView.findViewById(R.id.item_article_share);
        }
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/fave/favearticles/FaveArticlesAdapter.kt
package dev.ragnarok.fenrir.fragment.fave.favearticles

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import dev.ragnarok.fenrir.Constants
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.model.Article
import dev.ragnarok.fenrir.model.Photo
import dev.ragnarok.fenrir.nonNullNoEmpty
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.util.ViewUtils.displayAvatar

class FaveArticlesAdapter(private var data: List<Article>, private val context: Context) :
    RecyclerView.Adapter<FaveArticlesAdapter.Holder>() {
    private var clickListener: ClickListener? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Holder {
        return Holder(
            LayoutInflater.from(
                context
            ).inflate(R.layout.item_article, parent, false)
        )
    }

    override fun onBindViewHolder(holder: Holder, position: Int) {
        val article = data[position]
        holder.btFave.setImageResource(R.drawable.favorite)
        holder.btFave.setOnClickListener {
            clickListener?.onDelete(
                holder.bindingAdapterPosition,
                article
            )
        }
        if (article.uRL != null) {
            holder.ivButton.visibility = View.VISIBLE
            holder.ivButton.setOnClickListener {
                article.uRL.nonNullNoEmpty {
                    clickListener?.onArticleClick(article)
                }
            }
        } else holder.ivButton.visibility = View.GONE
        holder.btShare.setOnClickListener { clickListener?.onShare(article) }
        var photo_url: String? = null
        if (article.photo != null) {
            photo_url =
                article.photo?.getUrlForSize(Settings.get().main().prefPreviewImageSize, false)
        }
        if (photo_url != null) {
            holder.ivPhoto.visibility = View.VISIBLE
            displayAvatar(holder.ivPhoto, null, photo_url, Constants.PICASSO_TAG)
            holder.ivPhoto.setOnLongClickListener {
                article.photo?.let { it1 -> clickListener?.onPhotosOpen(it1) }
                true
            }
        } else holder.ivPhoto.visibility = View.GONE
        if (article.subTitle != null) {
            holder.ivSubTitle.visibility = View.VISIBLE
            holder.ivSubTitle.text = article.subTitle
        } else holder.ivSubTitle.visibility = View.GONE
        if (article.title != null) {
            holder.ivTitle.visibility = View.VISIBLE
            holder.ivTitle.text = article.title
        } else holder.ivTitle.visibility = View.GONE
        if (article.ownerName != null) {
            holder.ivName.visibility = View.VISIBLE
            holder.ivName.text = article.ownerName
        } else holder.ivName.visibility = View.GONE
    }

    override fun getItemCount(): Int {
        return data.size
    }

    fun setData(data: List<Article>) {
        this.data = data
        notifyDataSetChanged()
    }

    fun setClickListener(clickListener: ClickListener?) {
        this.clickListener = clickListener
    }

    interface ClickListener {
        fun onArticleClick(article: Article)
        fun onPhotosOpen(photo: Photo)
        fun onDelete(index: Int, article: Article)
        fun onShare(article: Article)
    }

    class Holder(root: View) : RecyclerView.ViewHolder(root) {
        val ivPhoto: ImageView = itemView.findViewById(R.id.item_article_image)
        val btFave: ImageView = itemView.findViewById(R.id.item_article_to_fave)
        val btShare: ImageView = itemView.findViewById(R.id.item_article_share)
        val ivSubTitle: TextView = itemView.findViewById(R.id.item_article_subtitle)
        val ivTitle: TextView = itemView.findViewById(R.id.item_article_title)
        val ivName: TextView = itemView.findViewById(R.id.item_article_name)
        val ivButton: Button = itemView.findViewById(R.id.item_article_read)
    }
}